### PR TITLE
Support parameter assignment before command

### LINKED ISSUE #138: Support `env` and parameter assignment before command
### Command name

env

### Completion context

```shell
# I know that the completion of git doesn't support GIT_DIR.
env GIT_EDITOR=vim GIT_DIR=/path/to git commit **<Tab>

# I know unimplemented, this is an example.
MANPAGER='nvim +Man!' man **<Tab>
```

### What happens now?

The completion of the command is shown while ignoring `env` and parameters.

### What do you expect to complete?

Suitable items for a command

### PR DESCRIPTION
**Pull request by hexium310**
This closes #138.

The arguments whose env is trimmed are passed to the function parsing arguments.
https://github.com/chitoku-k/fzf-zsh-completions/blob/3a90f2a89f37cab22a1d24f0f34a1152d9327fe0/src/completers/systemctl.zsh#L22

### CODE REVIEW

**Comment by hexium310** [comment]
@chitoku-k Could you advise me about the test? Is the writing method of the test with the `__fzf_extract_command` mock like it of `gh` (https://github.com/chitoku-k/fzf-zsh-completions/pull/145/files#diff-1e4d5b76e894df53f17522b23f286f5bf0f298045b12981baae32f4e8b9d8c8a) no problem?

**Comment by chitoku-k** [comment]
@hexium310 I didn't find any problems over mocking of gh in that commit; the argument assertions for `__fzf_extract_command` might be optional though.

File: tests/gh.zunit [hunk 1]
```diff
@@ -5,36 +5,79 @@
...
     _fzf_path_completion() {
         assert $# equals 2
         assert $1 same_as ''
         assert $2 same_as 'gh '
+
+        assert __fzf_extract_command mock_times 1
     }
 
     prefix=
     _fzf_complete_docker 'gh '
```
Comment (hexium310) [inline tests/gh.zunit:37]: I found the wrong function... In this file, there are some of them.

**Comment by chitoku-k** [comment]
What are they?

File: tests/gh.zunit (hunk 1 above)
Comment (chitoku-k) [reply tests/gh.zunit:37]: What are they?

Comment (hexium310) [reply tests/gh.zunit:37]: They are what I forgot to edit after copy-pasted.

**Comment by hexium310** [comment]
This is no matter, but I will paste the script of adding the mocks.

<details>
<summary>add_mock.vim</summary>

```vim
function! AddMock() abort
  call cursor([1, 1])
  while search("@test 'Testing completion.*:", 'W') > 0
    let function_end = search('^}$', 'nW')
    let command = matchlist(getline(search('_fzf_complete_.\+', 'nW', function_end)), '_fzf_complete_\k\+ \+\zs.\+$')
    call append(search('_fzf_complete()', 'nW') - 1, [
         \   '    __fzf_extract_command_mock_1() {',
         \   '        assert $# equals 1',
         \   '        assert $1 same_as ' . command[0],
         \   '',
         \   "        echo '" . expand('%:t:r') . "'",
         \   '    }',
         \   '',
         \ ])
    let function_end = search('^}$', 'nW')
    call search('_fzf_complete()', 'W', function_end)
    call cursor([line('.'), 0])
    if search('[[:space:]]*fail', 'nW', line('.') + 1) > 0
      if search('_fzf_path_completion', 'W', function_end) > 0
        call append(search('^    }$', 'W', function_end) - 1, [
              \   '',
              \   '        assert __fzf_extract_command mock_times 1',
              \ ])
      elseif search('_fzf_complete_\k\+ \+.\+$', 'W', function_end)
        call append(line('.'), [
              \   '',
              \   '    assert __fzf_extract_command mock_times 1'
              \ ])
        let function_end = search('^}$', 'nW')
        call search('assert __fzf_extract_command mock_times 1', 'W', function_end)
        if line('.') + 1 == search('^$', 'W', function_end)
          call deletebufline('%', line('.'))
        endif
      endif
    elseif search('run cat', 'W', function_end) > 0
      call append(line('.'), [
            \   '        assert __fzf_extract_command mock_times 1',
            \ ])
    endif
  endwhile
endfunction
```

</details>

File: src/completers/env.zsh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env zsh
+
+_fzf_complete_env() {
+    local arguments=(${(Q)${(z)@}})
```
Comment (chitoku-k) [inline src/completers/env.zsh:4]: Is this expansion safe?

File: src/completers/env.zsh [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env zsh
+
+_fzf_complete_env() {
+    local arguments=("${(Q)${(z)@}[@]}")
+    arguments=($(_fzf_complete_trim_env "${arguments[2,-1]}"))
```
Comment (chitoku-k) [inline src/completers/env.zsh:5]: Is this expansion safe again?

File: src/completers/systemctl.zsh [hunk 4]
```diff
@@ -17,8 +17,9 @@ _fzf_complete_systemctl-units() {
     local fzf_options=$1
     shift
 
+    local arguments=$(_fzf_complete_trim_env $@)
```
Comment (chitoku-k) [inline src/completers/systemctl.zsh:20]: How about changing this to the following for consistency?:
```suggestion
    local arguments=("${(Q)${(z)"$(_fzf_complete_trim_env $@)"}[@]}")
```

File: src/completers/env.zsh [hunk 5]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env zsh
+
+_fzf_complete_env() {
+    local arguments=("${(Q)${(z)@}[@]}")
+    arguments=($(_fzf_complete_trim_env "${${(q)arguments[2,-1][@]}}"))
+    local cmd=${(Q)arguments[1]}
+
+    if (( $+functions[_fzf_complete_$cmd] )); then
+        LBUFFER=${LBUFFER/env /}
+        _fzf_complete_$cmd ${@/env /}
```
Comment (chitoku-k) [inline src/completers/env.zsh:10]: Is this expansion safe again and again?

File: src/completers/env.zsh [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        LBUFFER="env $LBUFFER"
+        return
+    fi
+
+    _fzf_path_completion "$prefix" $@
```
Comment (chitoku-k) [inline src/completers/env.zsh:15]: `$@` without double quotations is unsafe.

File: src/completers/env.zsh (hunk 5 above)
Comment (hexium310) [reply src/completers/env.zsh:10]: Here is safe.

File: src/completers/env.zsh [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env zsh
+
+_fzf_complete_env() {
+    local arguments=("${(Q)${(z)@}[@]}")
+    arguments=($(_fzf_complete_trim_env "${${(q)arguments[2,-1][@]}}"))
```
Comment (chitoku-k) [inline src/completers/env.zsh:5]: Memo: this is safe 👍

File: src/completers/env.zsh (hunk 5 above)
Comment (chitoku-k) [reply src/completers/env.zsh:10]: I don't think so. Here is a function that simply prints the arguments delimited by a comma:
```zsh
print-args() {
    echo ${(pj:, :)@}
}

print-args '' 2 '' 4 # => ', 2, , 4'
```
When this function is called from a function like this, it looks like it drops empty strings:
```zsh
call-print-args() {
    print-args ${@/env /}
}

call-print-args '' 2 'env ' 4 # => '2, 4'
```

File: src/completers/git.zsh [hunk 8]
```diff
@@ -306,40 +306,40 @@ _fzf_complete_git() {
...
-                _fzf_complete_git-commit-messages '' $@
+                _fzf_complete_git-commit-messages '' "$@"
                 ;;
 
             --author|--date)
                 ;;
 
             -F|-t|--file|--pathspec-from-file|--template)
-                __fzf_generic_path_completion "${prefix#$prefix_option}" $@$prefix_option _fzf_compgen_path '' '' ' '
+                __fzf_generic_path_completion "${prefix#$prefix_option}" "$@"$prefix_option _fzf_compgen_path '' '' ' '
```
Comment (chitoku-k) [inline src/completers/git.zsh:320]:
```suggestion
                __fzf_generic_path_completion "${prefix#$prefix_option}" "$@$prefix_option" _fzf_compgen_path '' '' ' '
```

File: src/completers/git.zsh [hunk 9]
```diff
@@ -360,37 +360,37 @@ _fzf_complete_git() {
...
                 ;;
 
             --depth|--deepen|-j|--jobs|-o|--server-option|--shallow-since|--submodule-prefix|--upload-pack)
                 ;;
 
             *)
                 local repository
-                if [[ $@ =~ '--multiple' ]] || ! repository=$(_fzf_complete_parse_argument 3 1 "${${(q)arguments[@]}}" "${(F)git_options_argument_required}") && [[ -z $repository ]]; then
-                    _fzf_complete_git-repositories '--multi' $@
+                if [[ "$@" =~ '--multiple' ]] || ! repository=$(_fzf_complete_parse_argument 3 1 "${${(q)arguments[@]}}" "${(F)git_options_argument_required}") && [[ -z $repository ]]; then
```
Comment (chitoku-k) [inline src/completers/git.zsh:388]: It does not have to be like this.

File: src/completers/git.zsh [hunk 10]
```diff
@@ -670,33 +670,33 @@ _fzf_complete_git() {
...
-    _fzf_path_completion "$prefix" $@
+    _fzf_path_completion "$prefix" "$@"
 }
 
 _fzf_complete_git-commits() {
     local fzf_options=$1
     shift
 
-    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option$prefix_ref < <({
+    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option$prefix_ref < <({
```
Comment (chitoku-k) [inline src/completers/git.zsh:699]:
```suggestion
    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option$prefix_ref" < <({
```

File: src/completers/git.zsh [hunk 11]
```diff
@@ -727,7 +727,7 @@ _fzf_complete_git-commits-not-in-head() {
     local rev_list_head=$(git rev-list HEAD --oneline 2> /dev/null)
     rev_list_head=(${(q)${(f)rev_list_head}})
 
-    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option$prefix_ref < <(
+    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option$prefix_ref < <(
```
Comment (chitoku-k) [inline src/completers/git.zsh:730]:
```suggestion
    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option$prefix_ref" < <(
```

File: src/completers/git.zsh [hunk 12]
```diff
@@ -740,7 +740,7 @@ _fzf_complete_git-commit-messages() {
     local fzf_options=$1
     shift
 
-    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <(
+    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option < <(
```
Comment (chitoku-k) [inline src/completers/git.zsh:743]:
```suggestion
    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option" < <(
```

File: src/completers/git.zsh [hunk 13]
```diff
@@ -791,7 +791,7 @@ _fzf_complete_git-files_index() {
     local fzf_options=$2
     shift 2
 
-    _fzf_complete --ansi --read0 --print0 ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_ref < <(git ls-tree --name-only --full-tree -r -z ${(Z+n+)git_options} ${treeish-HEAD} 2> /dev/null)
+    _fzf_complete --ansi --read0 --print0 ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_ref < <(git ls-tree --name-only --full-tree -r -z ${(Z+n+)git_options} ${treeish-HEAD} 2> /dev/null)
```
Comment (chitoku-k) [inline src/completers/git.zsh:794]:
```suggestion
    _fzf_complete --ansi --read0 --print0 ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_ref" < <(git ls-tree --name-only --full-tree -r -z ${(Z+n+)git_options} ${treeish-HEAD} 2> /dev/null)
```

File: src/completers/git.zsh [hunk 14]
```diff
@@ -886,7 +886,7 @@ _fzf_complete_git-repositories() {
         groups=${(Q)${(F)${${(q)${(f)groups}}#remotes.}}}
     fi
 
-    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <({
+    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option < <({
```
Comment (chitoku-k) [inline src/completers/git.zsh:889]:
```suggestion
    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option" < <({
```

File: src/completers/git.zsh [hunk 15]
```diff
@@ -905,7 +905,7 @@ _fzf_complete_git-refs() {
     local fzf_options=$1
     shift
 
-    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option$prefix_ref < <(
+    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option$prefix_ref < <(
```
Comment (chitoku-k) [inline src/completers/git.zsh:908]:
```suggestion
    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option$prefix_ref" < <(
```

File: src/completers/git.zsh [hunk 16]
```diff
@@ -925,7 +925,7 @@ _fzf_complete_git-notes() {
...
```
Comment (chitoku-k) [inline src/completers/git.zsh:928]:
```suggestion
    _fzf_complete --ansi --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option$prefix_ref" < <(
```

File: src/completers/git.zsh [hunk 17]
```diff
@@ -939,7 +939,7 @@ _fzf_complete_git-show-files() {
     local fzf_options=$1
     shift
 
-    _fzf_complete --ansi --read0 --print0 ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <(
+    _fzf_complete --ansi --read0 --print0 ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option < <(
```
Comment (chitoku-k) [inline src/completers/git.zsh:942]:
```suggestion
    _fzf_complete --ansi --read0 --print0 ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option" < <(
```

File: src/completers/kubectl.zsh [hunk 18]
```diff
@@ -118,7 +118,7 @@ _fzf_complete_kubectl() {
     done
 
     subcommands=($(_fzf_complete_parse_argument 2 1 "${${(q)arguments[@]}}" "${(F)kubectl_options_argument_required}" || :))
-    namespace=$(_fzf_complete_parse_option_arguments '-n' '--namespace' "${(F)kubectl_options_argument_required}" $@$RBUFFER || :)
+    namespace=$(_fzf_complete_parse_option_arguments '-n' '--namespace' "${(F)kubectl_options_argument_required}" "$@"$RBUFFER || :)
```
Comment (chitoku-k) [inline src/completers/kubectl.zsh:121]:
```suggestion
    namespace=$(_fzf_complete_parse_option_arguments '-n' '--namespace' "${(F)kubectl_options_argument_required}" "$@$RBUFFER" || :)
```

File: src/completers/kubectl.zsh [hunk 19]
```diff
@@ -881,7 +881,7 @@ _fzf_complete_kubectl() {
             return
         fi
 
-        __fzf_generic_path_completion "${prefix#$prefix_option}" $@$prefix_option _fzf_compgen_path '' '' ' '
+        __fzf_generic_path_completion "${prefix#$prefix_option}" "$@"$prefix_option _fzf_compgen_path '' '' ' '
```
Comment (chitoku-k) [inline src/completers/kubectl.zsh:884]:
```suggestion
        __fzf_generic_path_completion "${prefix#$prefix_option}" "$@$prefix_option" _fzf_compgen_path '' '' ' '
```

File: src/completers/kubectl.zsh [hunk 20]
```diff
@@ -926,7 +926,7 @@ _fzf_complete_kubectl-resource-names() {
         kubectl_arguments+=(--all-namespaces)
     fi
 
-    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <(
+    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option < <(
```
Comment (chitoku-k) [inline src/completers/kubectl.zsh:929]:
```suggestion
    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option" < <(
```

File: src/completers/kubectl.zsh [hunk 21]
```diff
@@ -963,7 +963,7 @@ _fzf_complete_kubectl-containers() {
     local fzf_options=$1
     shift
 
-    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <({
+    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option < <({
```
Comment (chitoku-k) [inline src/completers/kubectl.zsh:966]:
```suggestion
    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option" < <({
```

File: src/completers/kubectl.zsh [hunk 22]
```diff
@@ -983,7 +983,7 @@ _fzf_complete_kubectl-ports() {
     local fzf_options=$1
     shift
 
-    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <(
+    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option < <(
```
Comment (chitoku-k) [inline src/completers/kubectl.zsh:986]:
```suggestion
    _fzf_complete --ansi --tiebreak=index --header-lines=1 ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option" < <(
```

File: src/completers/git.zsh (hunk 17 above)
Comment (hexium310) [reply src/completers/git.zsh:942]: I understood them. I will fix!

File: src/completers/kubectl.zsh [hunk 23]
```diff
@@ -998,7 +998,7 @@ _fzf_complete_kubectl-annotations() {
     local fzf_options=$1
     shift
 
-    _fzf_complete --ansi --read0 --print0 --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- $@$prefix_option < <({
+    _fzf_complete --ansi --read0 --print0 --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@"$prefix_option < <({
```
Comment (chitoku-k) [inline src/completers/kubectl.zsh:1001]:
```suggestion
    _fzf_complete --ansi --read0 --print0 --tiebreak=index ${(Q)${(Z+n+)fzf_options}} -- "$@$prefix_option" < <({
```

File: src/core/parser.zsh [hunk 24]
```diff
@@ -1,5 +1,12 @@
 #!/usr/bin/env zsh
 
+_fzf_complete_trim_env() {
+    local arguments=("${(Q)${(z)@}[@]}")
+    local cmd=$(__fzf_extract_command $@)
```
Comment (chitoku-k) [inline src/core/parser.zsh:5]:
```suggestion
    local cmd=$(__fzf_extract_command "$@")
```

File: tests/gh.zunit [hunk 25]
```diff
@@ -463,7 +594,7 @@
         '8  2nd OPEN PR  2nd-open  OPEN'
     )
 
-    run _fzf_complete_docker-images_post <<< ${(F)input}
+    run _fzf_complete_gh-pr_post <<< ${(F)input}
```
Comment (chitoku-k) [inline tests/gh.zunit:597]: Good catch 😭

**Review by chitoku-k — APPROVED** [review]
Appreciate this great work!

**Pull request merged**